### PR TITLE
server: create pod network as early as possible

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
when we're managing ns lifecycle, we don't need to wait until the infra container has started before setting up the networking resources, as we have all that is needed (the network namespace, and sandbox annotations).

A key to this is this allows the network create to happen before the infra container creation, which prioritizes the work in case of long running network creations.

this also fixes a case where we may fail to clean up the networking resources if we fail in marshalling the returned structure
#### Does this PR introduce a user-facing change?

```release-note
none
```

### CODE REVIEW

File: server/sandbox_run_linux.go [hunk 1]
```diff
@@ -698,6 +692,42 @@ func (s *Server) runPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
 		return nil, err
 	}
 
+	// now that we have the namespaces, we should create the network if we're managing namespace Lifecycle
+	var ips []string
+	var result cnitypes.Result
+
+	if s.config.ManageNSLifecycle {
+		ips, result, err = s.networkStart(ctx, sb)
```
Comment (saschagrunert) [inline server/sandbox_run_linux.go:700]: If it is time consuming, would it make sense to do it in a parallel go routine and join it at the end of the function?

Generally speaking, does a sandbox have to be ready immediately after the request happened? I would assume that the kubelet checks via the PLEG if the sandbox is ready in any case. This way we could do everything in the background (in theory).

**Review by saschagrunert — APPROVED** [review]
LGTM, just one point for general discussion.

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Review by mrunalp — APPROVED** [review]

**Comment by mrunalp** [comment]
This resulted in a regression:
```
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.302491982-07:00" level=info msg="About to check CNI network podman (type=bridge)"                                                                                                                                                                                              
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.306710373-07:00" level=info msg="Checking CNI network podman (config version=0.4.0)"                                                                                                                                                                                           
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.306730797-07:00" level=error msg="Error checking network: Interface veth7551742a Mac doesn't match: 0a:2a:e1:4b:31:c7 not found"                                                                                                                                               
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.306738236-07:00" level=error msg="Error while checking pod to CNI network \"podman\": Interface veth7551742a Mac doesn't match: 0a:2a:e1:4b:31:c7 not found"                                                                                                                   
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.306750625-07:00" level=info msg="NetworkStart: stopping network for sandbox 3ee7512fea24557c17419c99eccea17b4d31f21b16dd26409714ff0160e80b54" id=f0909f7d-5140-4d54-833f-e1a3ef2ec6b6 name=/runtime.v1alpha2.RuntimeService/RunPodSandbox                                      
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.313646630-07:00" level=info msg="Got pod network &{Name:kubernetes-dashboard-76ff8b5fbd-mmrl9 Namespace:kubernetes-dashboard ID:3ee7512fea24557c17419c99eccea17b4d31f21b16dd26409714ff0160e80b54 NetNS:/var/run/netns/7ffdce08-7c42-4668-b2ba-12dcd0830394 Networks:[{Name:podm
an Ifname:eth0}] RuntimeConfig:map[podman:{IP: MAC: PortMappings:[] Bandwidth:<nil> IpRanges:[]}]}"                                                                                                                                                                                                                                                         
Oct 05 16:59:49 localhost.localdomain crio[61418]: time="2020-10-05 16:59:49.313750930-07:00" level=info msg="About to del CNI network podman (type=bridge)"                                                                                                                                                                                                
Oct 05 16:59:49 localhost.localdomain crio[61418]: panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                                                                                  
Oct 05 16:59:49 localhost.localdomain crio[61418]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x88 pc=0x1a2eb4e]                                                                                                                                                                                                                                 
Oct 05 16:59:49 localhost.localdomain crio[61418]: goroutine 79 [running]:                                                                                                                                                                                                                                                                                  
Oct 05 16:59:49 localhost.localdomain crio[61418]: panic(0x1c9f720, 0x2e278b0)                                                                                                                                                                                                                                                                              
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /usr/local/go/src/runtime/panic.go:1064 +0x470 fp=0xc000573cc8 sp=0xc000573c10 pc=0x440890                                                                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]: runtime.panicmem(...)                                                                                                                                                                                                                                                                                    
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /usr/local/go/src/runtime/panic.go:212                                                                                                                                                                                                                                                           
Oct 05 16:59:49 localhost.localdomain crio[61418]: runtime.sigpanic()                                                                                                                                                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /usr/local/go/src/runtime/signal_unix.go:720 +0x413 fp=0xc000573cf8 sp=0xc000573cc8 pc=0x457593                                                                                                                                                                                                  
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/internal/oci.(*Container).Dir(...)                                                                                                                                                                                                                                                
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/internal/oci/container.go:336                                                                                                                                                                                                                                          
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/internal/lib/sandbox.(*Sandbox).createFileInInfraDir(0xc0002e25a0, 0x1f3111e, 0xf, 0xc000682f80, 0x33)                                                                                                                                                                            
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/internal/lib/sandbox/sandbox.go:351 +0x2e fp=0xc000573d70 sp=0xc000573cf8 pc=0x1a2eb4e                                                                                                                                                                                 
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/internal/lib/sandbox.(*Sandbox).SetNetworkStopped(0xc0002e25a0, 0x21fbe01, 0xc0002614a0, 0xc00081e240)                                                                                                                                                                            
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/internal/lib/sandbox/sandbox.go:342 +0x77 fp=0xc000573dc8 sp=0xc000573d70 pc=0x1a2ea77                                                                                                                                                                                 
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/server.(*Server).networkStop(0xc000394f00, 0x21fbec0, 0xc0002e7920, 0xc0002e25a0, 0x0, 0x0)                                                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/server/sandbox_network.go:158 +0x425 fp=0xc000573fa8 sp=0xc000573dc8 pc=0x1ab8625                                                                                                                                                                                      
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/server.(*Server).networkStart.func1(0xc0005743d0, 0x21fbf00, 0xc0007c8720, 0xc0002e25a0, 0xc000394f00, 0x21fbec0, 0xc0002e7920)                                                                                                                                                   
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/server/sandbox_network.go:41 +0xea fp=0xc000574000 sp=0xc000573fa8 pc=0x1ad5eea                                                                                                                                                                                        
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/server.(*Server).networkStart(0xc000394f00, 0x21fbf00, 0xc0007c8720, 0xc0002e25a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x21a6640, ...)                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/server/sandbox_network.go:58 +0x1187 fp=0xc000574388 sp=0xc000574000 pc=0x1ab7a67                                                                                                                                                                                      
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/server.(*Server).runPodSandbox(0xc000394f00, 0x21fbf00, 0xc0007c8720, 0xc000327a60, 0x0, 0x0, 0x0)                                                                                                                                                                                
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/server/sandbox_run_linux.go:700 +0x7855 fp=0xc000575820 sp=0xc000574388 pc=0x1ac4415                                                                                                                                                                                   
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/server.(*Server).RunPodSandbox(0xc000394f00, 0x21fbf00, 0xc0007c8720, 0xc000327a60, 0xc000394f00, 0x1, 0x1)                                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/server/sandbox_run.go:68 +0x49 fp=0xc000575868 sp=0xc000575820 pc=0x1aba369                                                                                                                                                                                            
Oct 05 16:59:49 localhost.localdomain crio[61418]: k8s.io/cri-api/pkg/apis/runtime/v1alpha2._RuntimeService_RunPodSandbox_Handler.func1(0x21fbf00, 0xc0007c8720, 0x1e3dca0, 0xc000327a60, 0xc00056f920, 0x1, 0x1, 0x130e848)                                                                                                                                
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.pb.go:7731 +0x8b fp=0xc0005758b0 sp=0xc000575868 pc=0x1352e0b                                                                                                                                                      
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/internal/log.UnaryInterceptor.func1(0x21fbf00, 0xc0007c84e0, 0x1e3dca0, 0xc000327a60, 0xc000327aa0, 0xc000327ac0, 0x4dce66, 0x5f7bb375, 0xc00e55ec47, 0x11a98612c0a1)                                                                                                             
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/internal/log/interceptors.go:56 +0x116 fp=0xc000575940 sp=0xc0005758b0 pc=0x13afff6                                                                                                                                                                                    
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x21fbf00, 0xc0007c84e0, 0x1e3dca0, 0xc000327a60, 0x20, 0x1d87fe0, 0xc000327a60, 0x1d27f40)                                                                                                                                      
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63 fp=0xc0005759a0 sp=0xc000575940 pc=0x1adc923                                                                                                                                                     
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/cri-o/cri-o/server/metrics.UnaryInterceptor.func1(0x21fbf00, 0xc0007c84e0, 0x1e3dca0, 0xc000327a60, 0xc000327aa0, 0xc000327ae0, 0x1adc9ba, 0x1d87fe0, 0xc000327b00, 0xc000327aa0)                                                                                                             
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/server/metrics/interceptors.go:24 +0xc6 fp=0xc000575a48 sp=0xc0005759a0 pc=0x1a4f0e6                                                                                                                                                                                   
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x21fbf00, 0xc0007c84e0, 0x1e3dca0, 0xc000327a60, 0xc0000d7f00, 0x0, 0xc000089b08, 0x417938)                                                                                                                                     
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63 fp=0xc000575aa8 sp=0xc000575a48 pc=0x1adc923                                                                                                                                                     
Oct 05 16:59:49 localhost.localdomain crio[61418]: github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x21fbf00, 0xc0007c84e0, 0x1e3dca0, 0xc000327a60, 0xc000327aa0, 0xc000327ac0, 0xc000089b78, 0x520186, 0x1dc8c00, 0xc0007c84e0)                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34 +0xd7 fp=0xc000575b18 sp=0xc000575aa8 pc=0x1adcb17                                                                                                                                                     
Oct 05 16:59:49 localhost.localdomain crio[61418]: k8s.io/cri-api/pkg/apis/runtime/v1alpha2._RuntimeService_RunPodSandbox_Handler(0x1f1a400, 0xc000394f00, 0x21fbf00, 0xc0007c84e0, 0xc0003a7920, 0xc0003ed230, 0x21fbf00, 0xc0007c84e0, 0xc000758380, 0x35c)                                                                                               
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.pb.go:7733 +0x150 fp=0xc000575b88 sp=0xc000575b18 pc=0x12b9310                                                                                                                                                     
Oct 05 16:59:49 localhost.localdomain crio[61418]: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000622600, 0x2213280, 0xc0008d2000, 0xc0000d7f00, 0xc00044bd10, 0x2e3dfb8, 0x0, 0x0, 0x0)                                                                                                                                                            
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/google.golang.org/grpc/server.go:1024 +0x522 fp=0xc000575e38 sp=0xc000575b88 pc=0x1281382                                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]: google.golang.org/grpc.(*Server).handleStream(0xc000622600, 0x2213280, 0xc0008d2000, 0xc0000d7f00, 0x0)                                                                                                                                                                                                  
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/google.golang.org/grpc/server.go:1313 +0xd34 fp=0xc000575f68 sp=0xc000575e38 pc=0x12853f4                                                                                                                                                                       
Oct 05 16:59:49 localhost.localdomain crio[61418]: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00022ef10, 0xc000622600, 0x2213280, 0xc0008d2000, 0xc0000d7f00)                                                                                                                                                                                 
Oct 05 16:59:49 localhost.localdomain crio[61418]:         /home/mrunalp/repos/cri-o/vendor/google.golang.org/grpc/server.go:722 +0xa5 fp=0xc000575fb8 sp=0xc000575f68 pc=0x1292245      
```

**Pull request merged**